src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/StoredProcedures/GetAvailableExportJobs/getAvailableExportJobs.js

```diff
@@ -0,0 +1,102 @@
+﻿/**
+* This stored procedure provides the following functionality:
+* - Retrieve list of available export jobs.
+* - The script determines list of available export jobs by looking at the list of running jobs as well as maximumNumberOfConcurrentJobsAllowed.
+* 
+* @constructor
+* @param {string} maximumNumberOfConcurrentJobsAllowedInString - The maximum number of concurrent jobs allowed in string.
+* @param {string} jobHeartbeatTimeoutThresholdInSecondsInString - The number of seconds allowed before the job is considered to be stale in string.
+*/
+
+function getAvailableExportJobs(maximumNumberOfConcurrentJobsAllowedInString, jobHeartbeatTimeoutThresholdInSecondsInString) {
```

Comment:
Can you explain why this is a stored procedure?

Reply:
We need a way to guarantee that maximumNumberOfConcurrentJobs is enforced regardless of how many instances of worker is running so to do that, this logic is implemented in the stored procedure.

Let's say we have 2 worker instances running but the maximumNumberOfConcurrentJobs is set to 1, we still want to make sure at any given point, we only have 1 job running actively. `CosmosDbDistributedLock` was another option but since we have all of the export job in the same partition, I used the stored procedure.

Reply:
But this sproc is just reading, so how is it preventing two instances from reading and therefore starting the same jobs simultaneously? Optimistic concurrency can solve that, but then the sproc would not be required.

Reply:
The stored procedure returns the number of available jobs based on the number of job that's currently running. For example, if there are 3 jobs queued and 1 job running, and the stored procedure is called with `maximumNumberOfConcurrentJobs = 1`, the stored procedure will return 0 available jobs.

Say the there is 2 job queued and 0 job running, with `maximumNumberOfConcurrentJobs = 1`, the stored procedure will return 1 job available. Let's assume there are 2 workers who both called the stored procedure at the same time, both of them get the job, and the first it tries to do is acquire the job by updating the status and optimistic concurrency will only allow one worker to update the job so we ensure that there is only one worker running.

If we don't use stored procedure, then the worker have to get list of running/queued jobs and then figure out which job is available to be picked up. Assume we have 3 instances of workers and the `maximumNumberOfConcurrentJobs = 2`. Each worker makes the query to get list of jobs (let's say job 1, 2, and 3). Let's say both worker A and worker B try to acquire job 1 at the same time. Worker A came in first and therefore were able to update the job record and worker B gets conflict. At this point, worker B can't just acquire job 2 because worker C might have acquired job 2 already so every time a worker gets a conflict trying to acquire the job, it needs to re-run the query to see if there is any job still available or not, which puts additional reads in the partition.

Maybe I am missing something. We can have quick chat about this one when you have time?

Reply:
Did you consider having the stored procedure query and reserve in one go? Kind of like a dequeue operation? (And it could optionally request up to n jobs at once).

Reply:
That makes sense. I will update the code.

Reply:
Updated the stored procedure to query and aquire the job at the same time.